x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/EventEmitterService.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.analytics;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.Marker;
+import org.apache.logging.log4j.MarkerManager;
+import org.elasticsearch.ResourceNotFoundException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.common.logging.ESLogMessage;
+
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.HashMap;
+import java.util.Map;
+
+public class EventEmitterService {
+    private static final Logger logger = LogManager.getLogger(EventEmitterService.class);
+    private static final Marker AUDIT_MARKER = MarkerManager.getMarker("org.elasticsearch.xpack.application.analytics");
+    private final AnalyticsCollectionResolver analyticsCollectionResolver;
+    private final ClusterState clusterState;
+
+    public EventEmitterService(AnalyticsCollectionResolver analyticsCollectionResolver, ClusterState clusterState) {
+        this.analyticsCollectionResolver = analyticsCollectionResolver;
+        this.clusterState = clusterState;
+    }
+
+    /**
+     * Emits an {@link AnalyticsEvent} to the appropriate collection.
+     *
+     * @param event The {@link AnalyticsEvent} to emit.
+     * @param listener The {@link ActionListener} to notify when the event has been emitted.
+     */
+    public void emitEvent(final AnalyticsEvent event, final ActionListener<AcknowledgedResponse> listener) {
+        final String collectionName = getAnalyticsEventCollectionName(event);
+        final AnalyticsCollection collection;
+
+        try {
+            collection = analyticsCollectionResolver.collection(clusterState, collectionName);
+        } catch (ResourceNotFoundException e) {
+            listener.onFailure(e);
+            return;
+        }
+
+        logger.info(AUDIT_MARKER, prepareFormattedEvent(collection, event));
+
+        listener.onResponse(AcknowledgedResponse.TRUE);
```

Comment:
nit: `AcknowledgedResponse` is generally for responses from the master node indicating the completion of a change to the cluster state, where `AcknowledgedResponse#FALSE` denotes that the cluster state was updated as requested but the change was not completely processed by all the nodes that should have processed it. It doesn't look like this applies here. If there are no circumstances where you expect to return `AcknowledgedResponse#FALSE` then generally `listener` should be an `ActionListener<Void>` instead, using `listener.onResponse(null);` to complete it successfully.

In cases where you need an empty response which `extends ActionResponse` then consider using `ActionResponse$Empty` instead, but note that it's generally best to limit the scope of this as much as possible, mapping in onto a `Void` as close to the network boundary as you can.

Reply:
Looks good now; one other nit: it's generally best to `git merge` the target branch instead of force-pushing to a PR branch - force-pushing it makes it somewhat harder to track the context of reviews as things evolve.